### ISSUE TITLE
[Feature] Sandboxed mode

### ISSUE THREAD

**Issue by 89luca89** [89luca89/distrobox#28]
Right now the distrobox's containers are created in privileged mode and share a lot of sensitive host's folder.

This is done because the aim is tight integration with the host, not sandboxing.

---

It would be nice to have an **optional** (see: disabled if not specified)  `--unprivileged` or a `--sandbox` flag in `distrobox-create` to have a more isolated container to work with.

**Comment by 89luca89** [comment]
Sandboxed distrobox should **not** integrate with the host as tightly as the normal one

`$HOME` should not be bind mounted
Host's rootfs should be mounted in readonly (or ideally not mounted at all)
Sockets should not be passed to the container

Sandboxed distrobox should **not** use `--privileged` and should **not** use `--ipc host, --pid host`

Thinking of functionality:

- `--sandbox` seems an understandable flag to use
- `--allow-path` could be used to allow some path to be mounted as read-write

**Comment by 89luca89** [comment]
Depends on #70

**Comment by BlauerHunger** [comment]
I think only implementing "fully sandboxed+unprivileged" and "fully integrated+privileged" modes are not the way to go, but rather adding options to configure it finer-grained like making only a specified set of devices available, only select sockets, mapping a custom selection of directories or other things possible for Flatpaks using Flatseal. 

Based on that you could then add presets for integrated vs. sandboxed.

**Comment by 89luca89** [comment]
Well there is a big difference between distrobox and Flatpak, distrobox is not an application distribution method.
On one hand I agree having some more fine grained support is good
On the other, it's and extreme work to add support to filter specific devices (like, usb pen1 passes, usb pen2 does not pass etc etc)

I was thinking more on the lines of

Sanbox mode:

1 - custom home
2 - you specify which path can the container access
3 - you decide if devices in /dev/ are accessible or not
4 - you decide if you share:
     - network namespace
     - process namespace
     - ipc namespace
     - mount namespace
     - (others)

And on that one can do a couple of profiles (low to high sandboxing)

**Comment by 89luca89** [comment]
Adding to before
we should give options to:

- mount in overlay, or not mount at all
    - /home
    - /
    - /dev
- disable additional group mappings: `run.oci.keep_original_groups=0`
- disable stuff like
    - network namespace
    - process namespace
    - ipc namespace
    - mount namespace
    - (others)

**Comment by ennec-e** [comment]
I had some ideas on sandboxing modes. 
This gives overview of the modes I thought of, describing general use cases and overall access to Host. 
The technical implementation specifics (what to mount / allow or exclude etc) of each mode shall be predetermined[?].

| Mode         | Short | **see**    | **touch**            | Comments                |
|--------------|-------|------------|----------------------|-------------------------|
| porous       | rw    | yes        | yes                  | (default currently)     |
| read-only    | ro    | yes        | nope                 |                         |
| overlay-fork | of    | touched up | self (~fantasy host) | ("just works" mode) [*] |
| isolate      | iso   | no         | no way!              | (fully locked down) [#] |                               |

[*]: try to be smoothe, smoother than porous even. does not conflict with sovereign Host. but [guest view] may diverge from ground truth (host state) 
[#]: as close to completely seperate and airgapped [guest, host] as possible

[*], [#]: >
            User should be able to **successfully** do **as guest** `sudo rm -rf /` (and deal with the consequences) without causing any hiccups on Host.
            In effect `overlay-fork` can be thought to be like `isolate` mode but the starting point is Host state (in -of), as opposed to a blank slate (in -iso).

[?]: User would make a simple choice of mode and distrobox should make intelligent decisions in the backend.

**Comment by 89luca89** [comment]
Great analysis yes, that is what I had in mind, with rw being the default.
For this to work we will need first to implement #119 and #120

And we will need a way to fix the `workdir` stuff (probaby depending on the level, we completely omit that?)

**Comment by bd4** [comment]
I was just thinking this, great to see it's already in the works!

Another possible use case here is running graphical windows apps with wine or native old / experimental proprietary applications, or even experimental open source apps. There is overlap with x11docker (and flatpak) here, so not sure what makes sense for distrobox; this use case requires more work re fine grained controls like sound, webcam, nested X server vs host X server, gpu vs not. Maybe makes sense to start simple with "isolate".

**Comment by 89luca89** [comment]
> I was just thinking this, great to see it's already in the works!
> 
> Another possible use case here is running graphical windows apps with wine or native old / experimental proprietary applications, or even experimental open source apps. There is overlap with x11docker (and flatpak) here, so not sure what makes sense for distrobox; this use case requires more work re fine grained controls like sound, webcam, nested X server vs host X server, gpu vs not. Maybe makes sense to start simple with "isolate".

Yea starting with a "simpler" sandbox mode

The target (at least generally speaking) is some sort of general isolation (not going to provide super-capillar controls like flatpak for example)
Adding such fine grained controls would be too much and out of scope of the project (imho), let's keep in mind that the point of distrobox (and toolbx) is **integration** not **isolation**, obviously if you need complete isolation and security it is faster adding just the bits you need to a plain podman/docker than removing all the integration from here :smile:

**Comment by bd4** [comment]
Yes simple is good, I think that is what makes toolbox/distrobox so great!

I think an interesting distinction with distrobox vs something like x11docker is mutability and persistence. x11docker and similar are designed to have an immutable core, that is changed by creating a new image, e.g. Dockerfile with new layers, and have mutable data exposed via volumes. Distrobox and toolbox on the other hand have a mutable core by default, you can just sudo dnf/apt/zypper etc as much as you want, so it's more like working in a standard chroot. I think both have their pros and cons. What excites me about distrobox+isolate is that it can provide a more chroot like workflow while still providing sandbox/isolation. I think LXC and systemd-nspawn are in this category as well, but so far only toolbox and distrobox hit the too simple not to use threshold (even easier than setting up schroot and debootstrap/pacstrap really).

The cool thing is, with just isolate + the ability to bind mount specific files (pass directly to docker/podman -v?), a lot of more advanced use cases should be possible, e.g. mostly isolated but expose gpu for CUDA/ROCm/oneAPI development. They may not all be trivial with a specialized switch, but possible and still easier than doing it manually with podman.

**Comment by 89luca89** [comment]
> Yes simple is good, I think that is what makes toolbox/distrobox so great!
> 
> I think an interesting distinction with distrobox vs something like x11docker is mutability and persistence. x11docker and similar are designed to have an immutable core, that is changed by creating a new image, e.g. Dockerfile with new layers, and have mutable data exposed via volumes. Distrobox and toolbox on the other hand have a mutable core by default, you can just sudo dnf/apt/zypper etc as much as you want, so it's more like working in a standard chroot. I think both have their pros and cons. What excites me about distrobox+isolate is that it can provide a more chroot like workflow while still providing sandbox/isolation. I think LXC and systemd-nspawn are in this category as well, but so far only toolbox and distrobox hit the too simple not to use threshold (even easier than setting up schroot and debootstrap/pacstrap really).
> 

Yes the point of toolbox and distrobox is simplicity and well use the big number of well maintained container images on the vaiorus registries :smile: 


> The cool thing is, with just isolate + the ability to bind mount specific files (pass directly to docker/podman -v?), a lot of more advanced use cases should be possible, e.g. mostly isolated but expose gpu for CUDA/ROCm/oneAPI development. They may not all be trivial with a specialized switch, but possible and still easier than doing it manually with podman.

As you can see in the docs here: 

- https://github.com/89luca89/distrobox/blob/main/docs/usage/distrobox-create.md
- https://github.com/89luca89/distrobox/blob/main/docs/usage/distrobox-enter.md

Distrobox already supports a `--additional-flags/-a` flag to pass directly the podman/docker flags to the container manager, in the docs the very example of adding volumes is reported :smile:

**Comment by leleleSDX** [comment]
would leveraging sandbox tools like firejail work for the moment?

**Comment by 89luca89** [comment]
> would leveraging sandbox tools like firejail work for the moment?

Didn't think of this, we could use bwrap maybe as it does not require root to run, but probably worth experimenting

**Comment by Aex12** [comment]
> Didn't think of this, we could use bwrap maybe as it does not require root to run, but probably worth experimenting

Doesn't podman already isolate the container from the host? Wouldn't it be redundant to bwrap podman?

**Comment by Meister1593** [comment]
> It's a small nit but I think the issue description is misleading or outdated:
> 
> > Right now the distrobox's containers are created in privileged mode
> 
> Currently podman boxes are unprivileged by default. They do have write access to host's home and other directories, though.

Podman containers are indeed unprivileged by default, but distrobox creates privileged containers, otherwise it wouldn't be able to access bunch of devices and stuff.

**Comment by Meister1593** [comment]
Any news on this?

**Comment by xerz-one** [comment]
I honestly need this so much I might as well make a PR if it's not expected in the short term

**Comment by xerz-one** [comment]
ok so while we're at it, here's a hacky solution I just crafted. This should be good enough for what most people expect I hope! Just declare `$NAME` and run the following

```
mkdir -p ~/.local/bin ~/.local/share/distrobox/$NAME
cp $(which distrobox)-{init,create,export,host-exec} ~/.local/bin/
sed -i '/--volume \/:\/run\/host:rslave/d' ~/.local/bin/distrobox-create
DBX_CONTAINER_CUSTOM_HOME=~/.local/share/distrobox/$NAME ~/.local/bin/distrobox-create --image alpine:latest --home ~/.local/share/distrobox/test --name $NAME
```

now you can just run it with `distrobox enter --no-workdir $NAME` :tada:

**Comment by 89luca89** [comment]
@xerz-one 

I'm not planning on working on this very soon

The plan anyway should be:

1 find what is needed to have the minimum possible to have working:
   - audio
   - video
   - xorg/wayland
   - internet/lan
2 create a "sandboxed" profile that just mounts the minimum

Anything more than that should not be in scope of the project, as the point of toolbox/distrobox type of software is the host's integration, and de-integrating too much would just be a waste of effort. If anyone needs further sandboxing, they can always use podman/docker directly

**Comment by FedX-sudo** [comment]
I was experimenting with this idea, and I was wondering if it would be benefitial to include a `--disalow-path` to compliment the `--allow-path`.

**Comment by 89luca89** [comment]
@xerz-one that's great finding thanks!
This will surely be useful in the future when I (or someone else? :eyes: ) will work on this

Thanks!

**Comment by TheComputerM** [comment]
Any updates on this?

**Comment by damianoognissanti** [comment]
Since the container only can see the home of the user running distrobox (but it sees it even when another home folder is specified for the container), isn't one solution that a sandbox-mode just creates another user without password (e.g. named distrobox) that runs distrobox-enter, so that when you run an exported program it will run something like:
"su -l distrobox -c distrobox-enter  -n debian -- /usr/bin/chromium %U"

**Comment by juhp** [comment]
> It turns out a more bare `podman` with [Pods](https://github.com/marhkb/pods) can easily achieve a sandboxed environment with support for desktop apps and nested desktop sessions. [..]
> Here's what's needed to get Podman to launch apps

@xerz-one, is the setup described in more detail somewhere?

**Comment by digitalsignalperson** [comment]
> Since the container only can see the home of the user running distrobox (but it sees it even when another home folder is specified for the container), isn't one solution that a sandbox-mode just creates another user without password (e.g. named distrobox) that runs distrobox-enter, so that when you run an exported program it will run something like: "su -l distrobox -c distrobox-enter -n debian -- /usr/bin/chromium %U"

This tool might help achieve this or be inspiration:  https://github.com/intgr/ego

**Comment by diegosouza** [comment]
My 2 cents to the discussion: some services (like Bitbucket Pipelines Cloud) do not support `--privileged`.
I was almost ready to deliver a reestructured pipeline and I could not finish because of this kind of limitation.

To sum up: `--unprivileged` could solve my problem.

**Comment by 89luca89** [comment]
Hi all
I did a basic draft that you can check here:

https://github.com/89luca89/distrobox/pull/818

Let me know how this works :+1:

**Comment by osalbahr** [comment]
This seems great for porting to macOS, actually (#36). On macOS I don’t need to interact with the base system from inside the distrobox, only access to dev tools (and the internet). So podman running in a VM shouldn’t be an issue for this mode.

**Comment by 89luca89** [comment]
I have just now merged #818 

As explained in the PR:

**THIS IS NOT A PROPER DATA SANDBOX**

This should ensure a basic unsharing between guest and hosts

This allows to:

-unshare-devsys: do not share host devices and sysfs dirs from host
-unshare-ipc: do not share ipc namemspace with host
-unshare-netns: do not share the net namespace with host
-unshare-process: do not share process namemspace with host

And an unshare-all to do all the above.

This allows to have only the minimal requirements to:

access user's HOME
launch applications with GUI/GPU/Audio/Video
Additional mountpoints can be declared with --volume

For now I'm not contemplating unsharing the home that would be a bit too much and denaturalise the purpose of distrobox itself.

**Comment by bigpod98** [comment]
i guess my suggestion https://github.com/89luca89/distrobox/issues/969 would fall around this

**Comment by anirbandey1** [comment]
Is there a way to atleast deny the container access to a few paths such as the host's ~/.gnupg , ~/.password-store, ~/.ssh ?

**Comment by Froggy232** [comment]
Hi,
First, thanks you a lot for your work!

I agree that a data sandbox mode with whitelist would be awesome, and I have the feeling that it would be really close to attain with options like custom home for container.
Is it really not an option to develop this feature in the future?
I find it sad that it got dismissed, it would really be useful for some containerized software (if not the majority of them) to not having access to home directory, and then having a whitelist system that permit to give access to one or few specifics folders. 
Of course, this feature would not be a default, it would just be an option at the creation of the container.

I think it would be really, really useful to a lot of people, myself included.
Anyway, thanks you a lot for your work, I use it everyday on universalblue and it's really amazing, thanks you again!

**Comment by IPlayZed** [comment]
> I think it would be really, really useful to a lot of people, myself included.

The point of this would be a platform-independent implemention of something like firejail for programs not made for the given platform of with sandboxing in mind, like a snap or flatpak package, as far as I see.

**Comment by damianoognissanti** [comment]
> Is there a way to atleast deny the container access to a few paths such as the host's ~/.gnupg , ~/.password-store, ~/.ssh ?

If you want to be as isolated as possible you can do this:

1) Create a user for your distroboxes (`-m` creates a home folder for the new user): `useradd -m distrobox`
2) Switch to the distrobox account with `su -l distrobox` (note that you want the `-l` flag, otherwise you will get permission errors).
3) Add `export DISPLAY=":0"` to your profile, either `/etc/profile` or `~/.profile`, preferably the latter, and source it with `source ~/.profile` to activate it without needing to close your terminal and redo step 2.
4) Run `distrobox-create` with whatever options you want and add `--unshare-all` if you wish to unshare as much as possible.
5) Enter with `distrobox-enter` and install everything you want.

If you wish to run a GUI application you can just run the following inside a terminal where you aren't inside your distrobox  
`xhost +si:localuser:distrobox`

I tried this with firefox and file:/// can just see the distrobox-user's home and not my regular home.

And you can of course populate your distrobox user's home with whatever files you wish to share with it (using `sudo cp ...` for example).

**Comment by IPlayZed** [comment]
@damianoognissanti The problem with this is that there is no easy way to share files between the logged in user and the distrobox dedicated user. (Maybe ACL for the desired directories, but that involves a lot of work and the needed permissions.) Also, the user needs to have permission to use sudo. (But it can be scoped to only distrobox, so I guess you could make the appropriate configs as you want in sudoers.)

**Comment by akhilman** [comment]
There is no problem hether to isolate a container from the host nor to
share selected parts of the host with the container. But this should be
a design decision made from the start and willingness to sacrifice some
ease of use.

Here is my prototype that I did to see if it is possible to have
isolation and the ability to run desktop applications at the same time.
https://codeberg.org/AkhIL/tinkerbox

One day I would like to rewrite this tool in Python, add cli
autocompletion, a GUI, and maybe even adding entries for isolated apps
in to the host's menu.

**Comment by 89luca89** [comment]
Reporting here the same comment I did in #1413 

> I'd like to reflect on the implication of this (and https://github.com/89luca89/distrobox/issues/28 )
> 
> What are the use cases for this?
> 
> The main use case of using distrobox over podman (or docker) is the integration it has with the system.
> And also the "ease of use", as you could do with podman what distrobox does, but distrobox abstracts lots of flags (so you can just distrobox enter and it works)
> 
> This invalidates all the usecases of distrobox
> 
> app/bin exporting
> host integration
> ease of use (we need to add all the various flags to use this)
> So my question is:
> 
> Isn't using podman directly a better alternative?
> 
> I feel like all of this PR (and https://github.com/89luca89/distrobox/issues/28 in general) are going in the opposite direction of what the projects aims to do.
> 
> Sandboxing is not a target for distrobox, it may add some unsharing bits for useful purposes (say, initful containers) but it's not aiming to replace podman/docker/flatpak

All in all I think a dedicated tool is more indicated than bending this project on something it is not designed to do.

**Comment by bigpod98** [comment]
better sandboxed mode would likely allow better user choice on what is allowed trough

**Comment by IPlayZed** [comment]
It is entirely up to the project maintainer if he does not want to implement stuff like this and let's be real, there already are several permission systems like ACL, capabilities etc., which allows separation. The problem is that it does not integrate with Distrobox and most people are not even aware how to use these and at the end of the day they should not be. I think that either there should be a new tool made, which makes it easier to separate Distrobox containers, as they are OCI containers or someone should present a PR for this, which can be considered as an example by @89luca89 . We should not expect him to go against what he thinks is best for this project.

**Comment by 89luca89** [comment]
> It is entirely up to the project maintainer if he does not want to implement stuff like this and let's be real, there already are several permission systems like ACL, capabilities etc., which allows separation. The problem is that it does not integrate with Distrobox and most people are not even aware how to use these and at the end of the day they should not be. I think that either there should be a new tool made, which makes it easier to separate Distrobox containers, as they are OCI containers or someone should present a PR for this, which can be considered as an example by @89luca89 . We should not expect him to go against what he thinks is best for this project.

Thanks @IPlayZed exaclty this

<SOME> isolation is useful for <SOME> functionalities (see netns for vpns, or procns for init) but this should not become a "let's implement all the flags to go back to a regular docker/podman"

For GUIs flatpaks do a better job at being "integrated but sandboxed", distrobox can be a stopgap for missing stuff, but it's not a replacement

And for sandboxed CLIs, regular docker/podman exists

I'm closing this as I don't want to implement further

**Comment by Gerharddc** [comment]
> > It is entirely up to the project maintainer if he does not want to implement stuff like this and let's be real, there already are several permission systems like ACL, capabilities etc., which allows separation. The problem is that it does not integrate with Distrobox and most people are not even aware how to use these and at the end of the day they should not be. I think that either there should be a new tool made, which makes it easier to separate Distrobox containers, as they are OCI containers or someone should present a PR for this, which can be considered as an example by [@89luca89](https://github.com/89luca89) . We should not expect him to go against what he thinks is best for this project.
> 
> Thanks [@IPlayZed](https://github.com/IPlayZed) exaclty this
> 
> isolation is useful for functionalities (see netns for vpns, or procns for init) but this should not become a "let's implement all the flags to go back to a regular docker/podman"
> 
> For GUIs flatpaks do a better job at being "integrated but sandboxed", distrobox can be a stopgap for missing stuff, but it's not a replacement
> 
> And for sandboxed CLIs, regular docker/podman exists
> 
> I'm closing this as I don't want to implement further

Perhaps a new tool is needed indeed, but I think that could be quite useful as I believe I have at least one use-case for a "sandboxed" version of Distrobox that I have not been able to find an alternative for after looking for two days now.

My use-case is that I would like to develop GUI apps inside a sandboxed container (that likely also contains my editor). Basically the only thing it should have access to is the source-code folder, the wayland socket, the GPU and networking. And some way of getting Git credentials would also be nice of course.

This is already mostly possible with VSCode but is difficult or impossible to achieve with other editors in my experience.

For "network only" app development it is easy enough to just use Podman directly and SSH into the container or possibly to use a tool like DevPod. But I have not been able to get this working for GUI apps whereas with Distrobox this all just works seamlessly (only without the sandboxing part of course).

I even tried running Firejail inside Distrobox to get this working but eventually gave up as it seems Firejail considers Distrobox to already be a sandbox and then just does nothing.

**Comment by stone-w4tch3r** [comment]
I added a little sandbox improvement in PR #1806
